onnxruntime/core/common/threadpool.cc

```diff
@@ -21,12 +21,222 @@ limitations under the License.
+#else
+    threal_pool_name_ = threal_pool_name;
+#endif
+  } else {
+    threal_pool_name_ = "unnamed_thread_pool";
+  }
+}
+
+ThreadPoolProfiler::~ThreadPoolProfiler() {
+  enabled_ = false;
+}
+
+void ThreadPoolProfiler::Start() {
+  enabled_ = true;
+}
+
+ThreadPoolProfiler::MainThreadStat& ThreadPoolProfiler::GetMainThreadStat() {
+  static thread_local std::unique_ptr<MainThreadStat> stat;
+  if (!stat) {
+    stat.reset(new MainThreadStat());
+  }
+  return *stat;
+}
+
+std::string ThreadPoolProfiler::Stop() {
+  ORT_ENFORCE(enabled_, "Profiler not started yet");
+  std::stringstream ss;
+  ss << "{\"main_thread\": {"
+     << "\"thread_pool_name\": \""
+     << threal_pool_name_ << "\", "
+     << GetMainThreadStat().Reset()
+     << "}, \"sub_threads\": {"
+     << DumpChildThreadStat()
+     << "}}";
+  return ss.str();
+}
+
+void ThreadPoolProfiler::LogCoreAndBlock(std::ptrdiff_t block_size) {
+  if (enabled_) {
+    GetMainThreadStat().LogCore();
+    GetMainThreadStat().LogBlockSize(block_size);
+  }
+}
+
+void ThreadPoolProfiler::LogStart() {
+  if (enabled_) {
+    GetMainThreadStat().LogStart();
+  }
+}
+
+void ThreadPoolProfiler::LogEnd(ThreadPoolEvent evt) {
+  if (enabled_) {
+    GetMainThreadStat().LogEnd(evt);
+  }
+}
+
+void ThreadPoolProfiler::LogEndAndStart(ThreadPoolEvent evt) {
+  if (enabled_) {
+    GetMainThreadStat().LogEndAndStart(evt);
+  }
+}
+
+void ThreadPoolProfiler::MainThreadStat::LogCore() {
+#ifdef _WIN32
+  core_ = GetCurrentProcessorNumber();
+#elif defined(__APPLE__)
+  uint32_t CPUInfo[4];
+  __cpuid_count(1, 0, CPUInfo[0], CPUInfo[1], CPUInfo[2], CPUInfo[3]);
+  if ((CPUInfo[3] & (1 << 9)) != 0) {
+    core_ = (unsigned)CPUInfo[1] >> 24;
+  }
+#else
+  core_ = sched_getcpu();
+#endif
+}
+
+void ThreadPoolProfiler::MainThreadStat::LogBlockSize(std::ptrdiff_t block_size) {
+  blocks_.emplace_back(block_size);
+}
+
+void ThreadPoolProfiler::MainThreadStat::LogStart() {
+  points_.emplace_back(Clock::now());
+}
+
+void ThreadPoolProfiler::MainThreadStat::LogEnd(ThreadPoolEvent evt) {
+  ORT_ENFORCE(!points_.empty(), "LogStart must pair with LogEnd");
+  events_[evt] += TimeDiffMicroSeconds(points_.back(), Clock::now());
+  points_.pop_back();
+}
+
+void ThreadPoolProfiler::MainThreadStat::LogEndAndStart(ThreadPoolEvent evt) {
+  ORT_ENFORCE(!points_.empty(), "LogStart must pair with LogEnd");
+  events_[evt] += TimeDiffMicroSeconds(points_.back(), Clock::now());
+  points_.back() = Clock::now();
+}
+
+std::string ThreadPoolProfiler::MainThreadStat::Reset() {
+  ORT_ENFORCE(points_.empty(), "LogStart must pair with LogEnd");
+  std::stringstream ss;
+  ss << "\"thread_id\": \"" << std::this_thread::get_id() << "\", \"block_size\": [";
+  if (!blocks_.empty()) {
+    std::copy(blocks_.begin(), blocks_.end() - 1, std::ostream_iterator<std::ptrdiff_t>(ss, ", "));
+    ss << blocks_.back();
+    blocks_.clear();
+  }
+  ss << "], \"core\": " << core_ << ", ";
```

Comment:
>core_ [](start = 27, length = 5)

should core_ be reset too?

Reply:
Yes, every time core_ will be reset.

Reply:
i meant that it doesn't seem to be written to in Reset(), unlike the other data members

---
In reply to: [597125546](https://github.com/microsoft/onnxruntime/pull/6748#discussion_r597125546) [](ancestors = 597125546)